File: src/Course.java
package src;

import java.util.ArrayList;
import java.util.Objects;

public class Course {
    private CourseInformation courseInformation;
    private CourseRequirements courseRequirements;
    private ArrayList<CourseSection> courseSections;

    public Course(CourseInformation courseInformation, CourseRequirements courseRequirements, ArrayList<CourseSection> courseSections) {
        this.courseInformation = courseInformation;
        this.courseRequirements = courseRequirements;
        this.courseSections = courseSections;
    }

    public String getCourseName() { // do not use this. Instead of use getter for courseInformation
        return getCourseInformation().getCourseName();
    }

    public CourseInformation getCourseInformation() {
        return courseInformation;
    }

    public CourseRequirements getCourseRequirements() {
        return courseRequirements;
    }

    public ArrayList<CourseSection> getCourseSections() {
        return courseSections;
    }
}

class CourseInformation {
    private final String courseName;
    private final String courseCode;

    CourseInformation(String courseName, String courseCode) {
        this.courseName = courseName;
        this.courseCode = courseCode;
    }

    public String getCourseCode() {
        return courseCode;
    }

    public String getCourseName() {
        return courseName;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        CourseInformation that = (CourseInformation) o;
        return Objects.equals(courseName, that.courseName) && Objects.equals(courseCode, that.courseCode);
    }

    @Override
    public int hashCode() {
        return Objects.hash(courseName, courseCode);
    }
}

class TakenCourse {
    private CourseInformation _courseInformation;
    private final int _midterm_score;
    private final int _final_score;
    private final Score _course_score;

    TakenCourse(CourseInformation courseInformation, int midtermScore, int finalScore) {
        _courseInformation = courseInformation;
        _midterm_score = midtermScore;
        _final_score = finalScore;
        _course_score = calculateCourseScore();
    }

    public CourseInformation get_courseInformation() {
        return _courseInformation;
    }


    private Score calculateCourseScore() {
        double total = _midterm_score * 0.6 + _final_score * 0.4;
        if (total >= 90) {
            return Score.AA;
        } else if (total >= 85) {
            return Score.BA;
        } else if (total >= 80) {
            return Score.BB;
        } else if (total >= 75) {
            return Score.CB;
        } else if (total >= 65) {
            return Score.CC;
        } else if (total >= 58) {
            return Score.DC;
        } else if (total >= 50) {
            return Score.DD;
        } else {
            return Score.FF;
        }
    }


    public Score get_course_score() {
        return _course_score;
    }

    public int get_midterm_score() {
        return _midterm_score;
    }

    public int get_final_score() {
        return _final_score;
    }
}

enum Score {
    AA,
    BA,
    BB,
    CB,
    CC,
    DC,
    DD,
    FF;
}

class CourseRequirements {
    private ArrayList<CourseInformation> _prerequisite_courses;
    private final int _minimum_current_class;
    private final DepartmentID _departmentID;
    private final FacultyID _facultyID;

    public CourseRequirements(ArrayList<CourseInformation> prerequisiteCourses, int minimumCurrentClass, FacultyID faculty, DepartmentID department) { // constructor
        _prerequisite_courses = prerequisiteCourses;
        _minimum_current_class = minimumCurrentClass;
        _departmentID = department;
        _facultyID = faculty;
    }

    public boolean[] isStudentQualified(Student student) {
        boolean[] ans = new boolean[4]; //Year, department, faculty, prerequisites
        ans[0] = student.getCurrentClass() >= _minimum_current_class; // check current class
        ans[1] = _departmentID == null || _departmentID.getDepartmentID() == student.get_studentID().get_departmentID().getDepartmentID(); // check department
        ans[2] = _facultyID == null || _facultyID.getFacultyID() == student.get_studentID().get_facultyID().getFacultyID(); //check faculty
        ans[3] = checkPrerequisiteCourse(student);                                  // check prerequisite courses
        return ans;
    }

    private boolean checkPrerequisiteCourse(Student student) {// check prerequisite courses method
        if (_prerequisite_courses == null || _prerequisite_courses.isEmpty()) {
            return true;
        }
        for (CourseInformation courseInformation : _prerequisite_courses) {
            boolean check = false;
            for (TakenCourse takenCourse : student.getTranscript().getTakenCourses()) {
                if (takenCourse.get_courseInformation().equals(courseInformation)) { //branchleri birlşetirirken takenCourse getterının ismini aynı yap
                    check = true;
                    break;
                }
            }
            if (!check) {
                return false;
            }
        }
        return true;
    }//

    public DepartmentID get_departmentID() {
        return _departmentID;
    }

    public FacultyID get_facultyID() {
        return _facultyID;
    }

    public int get_minimum_current_class() {
        return _minimum_current_class;
    }
}

class CourseSection {                                                        // CourseSection class
    public Day _day;
    public SectionTime _sectionTime;
    private Lecturer _lecturer;

    public CourseSection(Day day, SectionTime sectionTime, Lecturer lecturer) { // constructor
        _day = day;
        _sectionTime = sectionTime;
        _lecturer = lecturer;
    }

    boolean changeLecturer(Lecturer lecturer) {                              // change lecturer method
        _lecturer = lecturer;
        return true;
    }
}

File: pythonCode/Course.py
import Department
import Faculty
from Classroom import Classroom
from User import Lecturer
import logging

from pythonCode.Logger import setup_logger
logger = setup_logger("Course")

class Course:
    def __init__(self, courseInformation, courseRequirements, courseSections):
        logging.getLogger().handlers.clear()
        logger.info(f"{self.__class__.__name__} classes created.")
        self.courseInformation = courseInformation
        self.courseRequirements = courseRequirements
        self.courseSections = courseSections
        self._currentStudentCount = 0
        self._courseCapacity = int(1e9)
        for courseSection in courseSections:
            if courseSection._classRoom.get_capacity() < self._courseCapacity:
                self._courseCapacity = courseSection._classRoom.get_capacity()

    @staticmethod
    def from_dict(data):
        ci = CourseInformation.from_dict(data["courseInformation"])
        cr = CourseRequirements.from_dict(data["courseRequirements"])
        css = list()
        for data_cs in data["courseSections"]:
            css.append(CourseSection.from_dict(data_cs))

        return Course(ci, cr, css)

    def getCourseName(self):
        logger.info(f"The getCourseName method in the {self.__class__.__name__} class is called.")
        return self.courseInformation.getCourseName()

    def getCourseCode(self):
        logger.info(f"The getCourseCode method in the {self.__class__.__name__} class is called.")
        return self.courseInformation.getCourseCode()

    def getCourseInformation(self):
        logger.info(f"The getCourseInformation method in the {self.__class__.__name__} class is called.")
        return self.courseInformation

    def getCourseRequirements(self):
        logger.info(f"The getCourseRequirements method in the {self.__class__.__name__} class is called")
        return self.courseRequirements

    def getCourseSections(self):
        logger.info(f"The getCourseSections method in the {self.__class__.__name__} class is called.")
        return self.courseSections

    def getCourseCapacity(self):
        logger.info(f"The getCourseCapacity method in the {self.__class__.__name__} class is called.")
        return self._courseCapacity

    def getCurrentStudentCount(self):
        logger.info(f"The getCurrentStudentCount method in the {self.__class__.__name__} class is called.")
        return self._currentStudentCount

    def incrementCurrentStudentCount(self):
        logger.info(f"The incrementCurrentStudentCount method in the {self.__class__.__name__} class is called.")
        self._currentStudentCount += 1


class CourseInformation:
    def __init__(self, courseName, courseCode):
        logging.getLogger().handlers.clear()
        logger.info(f"{self.__class__.__name__} classes created.")
        self.courseName = courseName
        self.courseCode = courseCode

    @staticmethod
    def from_dict(data):
        return CourseInformation(data["courseName"], data["courseCode"])

    def getCourseCode(self):
        logger.info(f"The getCourseCode method in the {self.__class__.__name__} class is called.")
        return self.courseCode

    def getCourseName(self):
        logger.info(f"The getCourseName method in the {self.__class__.__name__} class is called.")
        return self.courseName


class TakenCourse:
    def __init__(self, courseInformation, midtermScore, finalScore):
        logging.getLogger().handlers.clear()
        logger.info(f"{self.__class__.__name__} classes created.")
        self._courseInformation = courseInformation
        self._midterm_score = midtermScore
        self._final_score = finalScore

    @staticmethod
    def from_dict(data):
        ci = CourseInformation.from_dict(data["_courseInformation"])
        ms = int(data["_midterm_score"])
        fs = int(data["_final_score"])
        return TakenCourse(ci, ms, fs)

    def getCourseInformation(self):
        logger.info(f"The getCourseInformation method in the {self.__class__.__name__} class is called.")
        return self._courseInformation

    def getMidtermScore(self):
        logger.info(f"The getMidtermScore method in the {self.__class__.__name__} class is called.")
        return self._midterm_score

    def getFinalScore(self):
        logger.info(f"The getFinalScore method in the {self.__class__.__name__} class is called.")
        return self._final_score

    def calculateCourseScore(self):
        logger.info("Calculating the course score.")
        total = self._midterm_score * 0.6 + self._final_score * 0.4
        if total >= 90:
            return Score.AA
        elif total >= 85:
            return Score.BA
        elif total >= 80:
            return Score.BB
        elif total >= 75:
            return Score.CB
        elif total >= 65:
            return Score.CC
        elif total >= 58:
            return Score.DC
        elif total >= 50:
            return Score.DD
        else:
            return Score.FF

    def getCourseScore(self):

        return self.calculateCourseScore()


class Score:
    AA = 4.0
    BA = 3.5
    BB = 3.0
    CB = 2.5
    CC = 2.0
    DC = 1.5
    DD = 1.0
    FF = 0.0


class CourseRequirements:
    def __init__(self, prerequisiteCourses, minimumCurrentClass, facultyID, departmentID):
        logging.getLogger().handlers.clear()
        logger.info(f"{self.__class__.__name__} classes created.")
        self._prerequisite_courses = prerequisiteCourses
        self._minimum_current_class = minimumCurrentClass
        self._facultyID = facultyID
        self._departmentID = departmentID

    @staticmethod
    def from_dict(data):
        prerequisite_courses_dict = data["_prerequisite_courses"]
        prerequisite_courses_list = list()

        for x in prerequisite_courses_dict:
            prerequisite_courses_list.append(CourseInformation.from_dict(x))

        mcc = int(data["_minimum_current_class"])
        fid = Faculty.FacultyID.from_dict(data["_facultyID"])
        did = Department.DepartmentID.from_dict(data["_departmentID"])

        return CourseRequirements(prerequisite_courses_list, mcc, fid, did)

    def isStudentQualified(self, student):
        logger.info("Checking if student is qualified for the course.")
        ans = [False, False, False, False]
        ans[0] = student.getCurrentClass() >= self._minimum_current_class
        ans[
            1] = self._departmentID is None or self._departmentID.getDepartmentID() == student.get_studentID().get_departmentID().getDepartmentID()
        ans[2] = self._facultyID is None or self._facultyID.getFacultyID() == student.get_studentID().get_facultyID().getFacultyID()
        ans[3] = self.checkPrerequisiteCourse(student)
        return ans

    def checkPrerequisiteCourse(self, student):
        logger.info("Checking if student has taken the prerequisite courses.")
        if self._prerequisite_courses is None or len(self._prerequisite_courses) == 0:
            return True

        for courseInformation in self._prerequisite_courses:
            check = False
            for takenCourse in student.getTranscript().getTakenCourses():
                if takenCourse.getCourseInformation().getCourseName()==courseInformation.getCourseName():
                    check = True
                    break

            if not check:
                return False

        return True

    def get_departmentID(self, _departmentID):
        logger.info(f"The get_departmentID method in the {self.__class__.__name__} class is called.")
        return self._departmentID

    def get_facultyID(self, _facultyID):
        logger.info(f"The get_facultyID method in the {self.__class__.__name__} class is called.")
        return self._facultyID

    def get_minimum_current_class(self):
        logger.info(f"The get_minimum_current_class method in the {self.__class__.__name__} class is called.")
        return self._minimum_current_class

    def getDepartmentID(self):
        logger.info(f"The getDepartmentID method in the {self.__class__.__name__} class is called.")
        return self._departmentID.getDepartmentID()

    def getFacultyID(self):
        logger.info(f"The getFacultyID method in the {self.__class__.__name__} class is called.")
        return self._facultyID.getFacultyID()


class CourseSection:
    def __init__(self, day, sectionTime, lecturer, classRoom):
        logging.getLogger().handlers.clear()
        logger.info(f"{self.__class__.__name__} classes created.")
        self._day = day
        self._sectionTime = sectionTime
        self._lecturer = lecturer
        self._classRoom = classRoom
        self._quota = classRoom.get_capacity()

    @staticmethod
    def from_dict(data):
        d = int(data["_day"])
        lec = Lecturer.from_dict(data["_lecturer"])
        sect = int(data["_sectionTime"])
        cr = Classroom.from_dict(data["_classRoom"])
        return CourseSection(d, sect, lec, cr)
